src/os/bluestore/BlueFSCache.cc

```diff
@@ -0,0 +1,203 @@
+    update_recency({inode, offset});
+    return 0;
+  }
+
+  // New insertion: check space
+  if (new_size > cache_capacity) {
+    dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " is greater than the cache capacity 0x" << cache_capacity << std::dec << dendl;
+    return -EINVAL;
+  }
+
+  // Evict until enough space
+  if (current_bytes + new_size > cache_capacity) {
+    dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " evict cache necessary" << std::dec << dendl;
+    evict(std::max(evict_bytes, cache_capacity));
+  }
+
+  dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " adding new entry to cache" << std::dec << dendl;
+  std::unique_lock lockIter(mutex_iter);
+  kv[{inode, offset}] = value;
+  access_order.push_front({inode, offset});
+  key_iter[{inode, offset}] = access_order.begin();
+  current_bytes += new_size;
+  dout(20) << __func__ << " current cache size 0x" << std::hex << current_bytes << " and cache capacity 0x" << cache_capacity << std::dec << dendl;
+  return 0;
+}
+
+void LRUCache::remove(std::pair<uint64_t, uint64_t> key) {
+  // Move key to front (most recently used)
+
+  dout(20) << __func__ << " removing key from cache for inode " << key.first << std::hex << " 0x" << key.second << std::dec << dendl;
+  {
+    std::unique_lock lock(mutex);
+    auto kv_it = kv.find(key);
+    if (kv_it != kv.end()) {
+      kv_it->second->clear();
+      kv.erase(key);
+      dout(20) << __func__ << " removed key from cache for inode " << key.first << std::hex << " 0x" << key.second << std::dec << dendl;
+    }
+  }
+
+  {
+    std::unique_lock lock_iter(mutex_iter);
+    auto list_it = key_iter.find(key);
+    if (list_it != key_iter.end()) {
+      key_iter.erase(key);
+      access_order.remove(key);
+      dout(20) << __func__ << " removed key from LRU list " << key.first << std::hex << " 0x" << key.second << std::dec << dendl;
+    }
+  }
+}
+
+void LRUCache::print_cache() {
+  dout(30) << __func__ << "Cache Contents (Total Bytes: 0x" << std::hex << current_bytes << std::dec << ")" << dendl;
+  for (auto key : access_order) {
+    auto &val = kv[key];
+    dout(30) << __func__ << std::hex << " inode " << key.first << " and offset 0x" << key.second << std::dec << dendl;
+  }
+}
+
+void LRUCache::update_recency(std::pair<uint64_t, uint64_t> key) {
+  // Move key to front (most recently used)
+
+  dout(20) << __func__ << " updating recency for inode " << key.first << " and offset 0x" << std::hex << key.second << std::dec << dendl;
+  std::unique_lock lock(mutex_iter);
+  auto it = key_iter.find(key);
+  if (it != key_iter.end()) {
+    access_order.erase(key_iter[key]);
+  }
+  access_order.push_front(key);
+  key_iter[key] = access_order.begin();
+}
+
+void LRUCache::evict(size_t target_free_bytes) {
```

Comment:
That is not clear to me if the function has to get rid of at least `target_free_bytes` capacity,
or after function is done, there must be at least `target_free_bytes` available.